lib/src/backend.rs

```diff
@@ -169,18 +169,42 @@ content_hash! {
 
 #[derive(Debug, Error)]
 pub enum BackendError {
-    #[error(
-        "Invalid hash for object of type {object_type}: {hash} (expected {expected} bytes, got \
-         {actual} bytes)"
-    )]
+    #[error("Invalid {id_type}: {hash} (expected {expected} bytes, got {actual} bytes)")]
     InvalidHashLength {
         expected: usize,
         actual: usize,
-        object_type: &'static str,
+        id_type: &'static str,
+        hash: String,
+    },
+    #[error("Invalid {id_type}: {hash}: {source}")]
+    InvalidHash {
+        id_type: &'static str,
         hash: String,
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+    #[error("Invalid UTF-8 for {id_type} {hash}: {source}")]
+    InvalidUtf8 {
+        id_type: &'static str,
+        hash: String,
+        source: std::string::FromUtf8Error,
+    },
+    #[error("Object with {id_type} {hash} not found: {source}")]
+    ObjectNotFound {
+        id_type: &'static str,
+        hash: String,
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+    #[error("Error when reading object with {id_type} {hash}: {source}")]
+    ReadObject {
+        id_type: &'static str,
+        hash: String,
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+    #[error("Could not write {object_type}: {source}")]
+    WriteObject {
+        object_type: &'static str,
```

Comment:
nit: s/object_type/id_type/? Or do the opposite rename above. I think I slightly prefer `object_type` to `id_type`, but you can decide.